### DISCUSSION TITLE
[Item] all annotations and notes by tag

### DISCUSSION THREAD

**Discussion by windingwind** [windingwind/zotero-better-notes#241] [Note Templates]
## Description

Author: @windingwind  
Discuss: #191  

> Only for Better Notes 1.1.3+

> Search annotations & notes with these tags in the selected items.

- Tag 1
    - note link 1
    - annotation 1
    - annotation 2
- Tag 2
    - note link 2
- Tag 3
    - annotation 3
 
![image](https://user-images.githubusercontent.com/33902321/198633450-4138f35e-60d5-41f3-9488-8e52475ac4cc.png)

## Import Template

One-click to import.

1. Copy the template below (just an example).
<details style="text-indent: 2em">
<summary>Click to show the example template</summary>

```yaml
name: "[Item] all annotations and notes by tag"
content: |-
  // @beforeloop-begin
  ${{
    sharedObj.tagsDict = {};
    sharedObj.updateData = (item, tag) => {
      if (!(tag in sharedObj.tagsDict)) {
        sharedObj.tagsDict[tag] = [];
      }
      sharedObj.tagsDict[tag].push(item);
    };
    return "";
  }}$
  // @beforeloop-end
  // @default-begin
  ${{
    const annotationsAndNotes = [];
    if (topItem.isRegularItem()) {
      const attachments = Zotero.Items.get(topItem.getAttachments()).filter((i) =>
        i.isPDFAttachment()
      );
      attachments.forEach((i) => {
        annotationsAndNotes.splice(0, 0, ...i.getAnnotations());
      });
      const notes = Zotero.Items.get(topItem.getNotes());
      annotationsAndNotes.splice(0, 0, ...notes);
    } else if (topItem.isNote()) {
      annotationsAndNotes.push(topItem);
    }
    annotationsAndNotes.forEach((item) =>
      item
        .getTags()
        .map((tag) => tag.tag)
        .forEach((tag) => {
          sharedObj.updateData(item, tag);
        })
    );
    return "";
  }}$
  // @default-end
  // @afterloop-begin
  ${{
    let res = "";
    for (const tag in sharedObj.tagsDict) {
      if(tag.startsWith('zotero://')){
        continue;
      }
      res += `<h1>Tag: ${tag}</h1>\n`;
      for (const note of sharedObj.tagsDict[tag].filter((item) =>
        item.isNote()
      )) {
        res += `<p><a href="${Zotero.BetterNotes.api.convert.note2link(
          note
        )}">${note.getNoteTitle()}</a></p>\n`;
      }
      res += await Zotero.BetterNotes.api.convert.annotations2html(
        sharedObj.tagsDict[tag].filter((item) => item.isAnnotation()),
        {
          noteItem: targetNoteItem
        }
      );
      res += "\n";
    }
    return res;
  }}$
  // @afterloop-end
```

</details>

2. Goto Zotero menubar, click `Edit`->`New Template from Clipboard`.
3. Click OK.

**Comment by Ravenclawer** [comment]
I keep getting this massage when applying this template
![image](https://github.com/windingwind/zotero-better-notes/assets/73892411/014b8660-27fe-49b4-a12f-f944c1c81e54)

**Reply by anandkumar89** [reply]
I'm also facing same problem while interesting template on workspace note.

**Reply by windingwind** [reply]
fixed. please import the template again.

**Comment by Ravenclawer** [comment]
now it keeps showing another error for every item 
![image](https://github.com/windingwind/zotero-better-notes/assets/73892411/576a8df8-d49d-417d-84db-22e8bf8ca930)

**Reply by windingwind** [reply]
Only for 1.1.3+, which only runs on Zotero 7.
Zotero 6 versions (<1.1.0) are no longer maintained.

**Reply by windingwind** [reply]
And this is a format for template import. Do not directly paste into the template editor. Please read the instructions.

**Reply by Ravenclawer** [reply]
OK, zotero7 is still under development， so I may stay at zotero6 for a while。

**Comment by jppmarujo** [comment]
Just what i've been looking for! You just made writing my thesis a whole lot easier, @windingwind, so thank you very much for being **AWESOME**.

**Reply by Rhydderch** [reply]
Hi @jppmarujo ! I'm working on my thesis as well and I just discovered better notes; could you explain a bit your workflow with better notes?

**Comment by logicfollows** [comment]
I've been trying to troubleshoot this feature for so many days! I know it is not longer maintained in Zotero 6 so just to use this I finally just updated to Zotero 7 and used your template and it finally worked!! Thank you so much!

**Reply by dexter-gh** [reply]
Did you change anything else? I just updated to Zotero 7 but still get the exact same error (SyntaxError: missing : after property id SyntaxError: missing : after property id SyntaxError: missing : after property id)... Any tips?

**Comment by Lyricism0001** [comment]
Hi @windingwind, thank you for your amazing work! 
Is there a way to get the results with the tags in alphabetical order? Or, if not, is there a way to reorder them alphabetically after the note is obtained? The pictures shows how I’m currently getting the results. The second picture is the tree view.
Thank you so much!
![image](https://github.com/user-attachments/assets/693415bd-5526-4beb-83d5-ae1853de9e5e)
![image](https://github.com/user-attachments/assets/c64c6461-d83e-4f11-bbc5-eddc4751428b)

**Comment by gregorylearns** [comment]
Still works in July 26,2025 with Zotero 7.0.22 (binary) in Linux using BetterNotes 2.5.3
